src/Simplify/AstHelpers.elm

```diff
@@ -647,16 +653,208 @@ getReducedLambda expressionNode =
+
+                        matchesRecordUpdate : Bool
+                        matchesRecordUpdate =
+                            -- checking \({x} as r) -> {r|x=x} match
+                            case unparenthesizedAliasedPattern of
+                                Node _ (Pattern.RecordPattern patternFields) ->
+                                    case removeParens expressionNode of
+                                        Node _ (Expression.RecordUpdateExpression (Node _ updatedRecordReferenceName) expressionFields) ->
+                                            (patternAliasName == updatedRecordReferenceName)
+                                                && list2AreSameLengthAndAll
+                                                    (\(Node _ ( Node _ expressionFieldName, expressionFieldValueNode )) (Node _ patternFieldName) ->
+                                                        (patternFieldName == expressionFieldName)
+                                                            && (case removeParens expressionFieldValueNode of
+                                                                    Node _ (Expression.FunctionOrValue [] expressionFieldValueReferenceName) ->
+                                                                        patternFieldName == expressionFieldValueReferenceName
+
+                                                                    _ ->
+                                                                        False
+                                                               )
+                                                    )
+                                                    (expressionFields
+                                                        |> List.sortBy (\(Node _ ( Node _ fieldName, _ )) -> fieldName)
+                                                    )
+                                                    (patternFields
+                                                        |> List.sortBy Node.value
+                                                    )
+
+                                        _ ->
+                                            False
+
+                                _ ->
+                                    False
+                    in
+                    matchesRecordUpdate
+                        || expressionReconstructsIrrefutablePattern lookupTable
+                            unparenthesizedExpressionNotLocalReference
+                            unparenthesizedAliasedPattern
+
+        Pattern.TuplePattern [ patternPart0, patternPart1 ] ->
+            case getTuple2 lookupTable expressionNode of
+                Nothing ->
+                    False
+
+                Just expressionParts ->
+                    expressionReconstructsIrrefutablePattern lookupTable expressionParts.first patternPart0
+                        && expressionReconstructsIrrefutablePattern lookupTable expressionParts.second patternPart1
```

Comment:
Similarly, this would enable a bit more TCO. I don't know how much we want to rewrite code to make this more TCO, as it does make things a bit less readable.
```suggestion
                    if expressionReconstructsIrrefutablePattern lookupTable expressionParts.first patternPart0 then
                        True

                    else
                        expressionReconstructsIrrefutablePattern lookupTable expressionParts.second patternPart1
```